Allow Re-Packager to split packages with multiple crafting recipes into single packages (#7439, #7484) (1.20.1)

AsterAether opened the pull request:
Adds the ability for the Re-Packager to split a package (created by requesting recipes via the JEI integration) into single packages, one for each "craft", to allow a setup where a mechanical crafter can process such requests.

Should fix #7484 and #7439.

## Implementation details
Adds an additional tag to `PackageOrderRequestPacket` and `PackageItem`, a `PackageOrderCraftingContext`, which is used to carry recipe information (item placement, and amount of requested recipe) instead of the previous context.

Previous context is now only used for defragmenting a fragmented package order.

RedstoneRequester also uses this field to request correct placement in crafter.

## Minor changes
Re-Packager now uses the `queuedExitingPackages` inventory to deposit split packages and overflow from defragmenting.
Re-Packager now uses the same `buttonCooldown` as Packager.

## Notes
This will break "in-flight" crafting packages of older versions, as their context now only is used for defragging. Is that alright / should that be handled?

AsterAether committed:
Allow Re-Packager to split crafting packet with multiple crafting requests into multiple packages with only one request per package.

Major:
Adds ability for package orders to contain multiple order contexts, with a requested amount.
Add ability to split packages on these context to Re-Packager to allow valid crafting packages to be output.

Minor:
Re-Packager now uses internal queuedExitingPackages to output split packages and overflow from fragmented.
Re-Packager also has button-delay now for redstone activations.

AsterAether committed:
Merge remote-tracking branch 'upstream/mc1.20.1/dev' into mc1.20.1/repackager-splitting

# Conflicts:
#	src/main/java/com/simibubi/create/content/logistics/packager/PackagerBlockEntity.java

AsterAether committed:
Update for UnpackingHandler

Fix packet not being removed after splitting.

AsterAether committed:
Merge branch 'mc1.20.1/dev' into mc1.20.1/repackager-splitting

# Conflicts:
#	src/main/java/com/simibubi/create/content/logistics/packagerLink/LogisticallyLinkedBehaviour.java
#	src/main/java/com/simibubi/create/content/logistics/packagerLink/LogisticsManager.java
#	src/main/java/com/simibubi/create/content/logistics/packagerLink/PackagerLinkBlockEntity.java
#	src/main/java/com/simibubi/create/content/logistics/stockTicker/StockCheckingBlockEntity.java
#	src/main/java/com/simibubi/create/content/logistics/stockTicker/StockTickerBlockEntity.java
#	src/main/java/com/simibubi/create/impl/unpacking/CrafterUnpackingHandler.java

AsterAether committed:
Do not replace context, instead add new craftingContext to PackageOrderRequestPacket and PackageItem.

AsterAether committed:
JavaDoc Fix

AsterAether committed:
Merge branch 'mc1.20.1/dev' into mc1.20.1/repackager-splitting

AsterAether commented:
I got the 1.21.1 port for this done as well, If you decide to merge, should I create a second PR then?

HttpRafa commented:
I think this fixes https://github.com/Creators-of-Create/Create/issues/7576 too

UrBoiTom commented:
Can't wait for this to be merged

simibubi committed:
Patching it up

- Fixed crash when repackaging non-crafting packages
- Fixed typo in nbt guard
- Inverted logic in buttonCooldown return
- Fixed re-packager able to overwrite held box while still exporting previously split packages

PokerfaceCactus commented:
would this allow gauges that utilize mechanical crafting to be set to higher output/craft counts too or is this exclusively for when using a ticker to send packagers to re-packagers?

simibubi committed:
Just a few tweaks

- Refactor away some of the data redundancy in package requesting methods and package nbt
- Fixed defragmenter not splitting by recipe when not all contents of combined requests fits into one package
- Packagers no longer store queued packages as individual instances if they are identical

simibubi commented:
Thanks for the pr!
We can take care of the 1.21 port for this internally
